modules/charon-core/src/test/java/org/wso2/charon3/core/utils/PatchOperationUtilTest.java

```diff
@@ -0,0 +1,400 @@
+        User oldResource = getUser();
+        User copyOfOldResource = (User) CopyUtil.deepCopy(oldResource);
+        SCIMResourceTypeSchema scimResourceTypeSchema = getSchema();
+
+        return new Object[][]{
+
+                {patchOperation1, jsonDecoder, oldResource, copyOfOldResource, scimResourceTypeSchema},
+                {patchOperation2, jsonDecoder, oldResource, copyOfOldResource, scimResourceTypeSchema},
+        };
+    }
+
+    @Test(dataProvider = "dataForPatchAddExceptions",
+            expectedExceptions = {NotImplementedException.class, BadRequestException.class})
+    public void testDoPatchAddExceptions(PatchOperation operation, JSONDecoder decoder, AbstractSCIMObject oldResource,
+                                         AbstractSCIMObject copyOfOldResource, SCIMResourceTypeSchema schema)
+            throws CharonException, BadRequestException, NotImplementedException, InternalErrorException {
+
+        PatchOperationUtil.doPatchAdd(operation, decoder, oldResource, copyOfOldResource, schema);
+
+    }
+
+    @DataProvider(name = "dataForPatchReplaceSuccess")
+    public Object[][] dataToPatchReplaceSuccess() throws InstantiationException, IllegalAccessException,
+            CharonException {
+
+        Map<String, String> operationValueName = new HashMap<>();
+        operationValueName.put("givenName", "John");
+        operationValueName.put("familyName", "Anderson");
+
+        PatchOperation patchOperation1 = PatchOperation.class.newInstance();
+        PatchOperation patchOperation2 = PatchOperation.class.newInstance();
+
+        patchOperation1.setOperation(replace);
+        patchOperation1.setPath("manager.displayName");
+        patchOperation1.setValues("Rash");
+
+        patchOperation2.setOperation(replace);
+        patchOperation2.setPath("emails[type eq home].value");
+        patchOperation2.setValues("rash123@gmail.com");
+
+        JSONDecoder jsonDecoder = new JSONDecoder();
+        User oldResource = getUser();
+        User copyOfOldResource = (User) CopyUtil.deepCopy(oldResource);
+        SCIMResourceTypeSchema scimResourceTypeSchema = getSchema();
+
+        return new Object[][]{
+
+                {patchOperation1, jsonDecoder, oldResource, copyOfOldResource, scimResourceTypeSchema},
+                {patchOperation2, jsonDecoder, oldResource, copyOfOldResource, scimResourceTypeSchema}
+        };
+    }
+
+    @Test(dataProvider = "dataForPatchReplaceSuccess")
+    public void testDoPatchReplaceSuccess(PatchOperation operation, JSONDecoder decoder, AbstractSCIMObject oldResource,
+                                          AbstractSCIMObject copyOfOldResource, SCIMResourceTypeSchema schema)
+            throws CharonException, NotImplementedException, BadRequestException, InternalErrorException {
+
+        PatchOperationUtil.doPatchReplace(operation, decoder, oldResource, copyOfOldResource, schema);
+
+        Assert.assertTrue(true, "doPatchReplace is successful");
```

Comment:
Don't we need to do any validation?

Reply:
fixed in 6e37f6e580f0e7c973edbc994d55b5a8062d5c14